llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp

```diff
@@ -2643,8 +2643,12 @@ bool IRTranslator::translateCall(const User &U, MachineIRBuilder &MIRBuilder) {
       MPI = MachinePointerInfo(Info.ptrVal, Info.offset);
     else if (Info.fallbackAddressSpace)
       MPI = MachinePointerInfo(*Info.fallbackAddressSpace);
-    MIB.addMemOperand(
-        MF->getMachineMemOperand(MPI, Info.flags, MemTy, Alignment, CI.getAAMetadata()));
+
+    if (Info.MMO)
+      MIB.addMemOperand(Info.MMO);
+    else
+      MIB.addMemOperand(MF->getMachineMemOperand(
+          MPI, Info.flags, MemTy, Alignment, CI.getAAMetadata()));
```

Comment:
I think this is an awkward API; either the IntrinsicInfo struct should contain all of the fields necessary to construct the MMO, or every instance of getTgtMemIntrinsic should be required to create the MMO.

Reply:
I think it would be too much too change every instance of getTgtMemIntrinsic in one PR, but that could be the end goal.

Reply:
Don't you just need to add ordering/syncscope fields?

Reply:
Yes, but @nhaehnle suggested to do this refactoring.

Reply:
It's not just syncscope and ordering, but also failure ordering, AAInfo and Ranges. Basically, you end up replicating MachineMemOperand in a second place, which I find pretty questionable.

I agree that the desirable end result is that everybody just creates `MachineMemOperand`s. Adding the field in IntrinsicInfo in this way was my proposal to provide a smoother upgrade path, but perhaps an alternative can be found.

For example, perhaps a second overload
```
virtual MachineMemOperand *getTgtMemIntrinsic(const CallInst &Call, MachineFunction &MF, unsigned IntrinsicID) 
```
could be added whose default implementation forwards to the currently-existing overload.

Reply:
I think there's value in fully decoupling this from the Machine infrastructure though; in principle we should be able to perform IR optimizations with intrinsic knowledge prior to codegen

Reply:
Fair enough. I don't care *that* much about this.